More like isearch

andyleejordan opened the pull request:
This implements most of the things I felt were missing in #47:

- `isearch-forward-symbol-at-point`
- `isearch-occur`
- isearch-like `M-s` bindings for those and existing ctrlf functions

The notable exception is that I haven't implemented `C-w` to yank the word at point into the search. This also needs changelog updates (and tests?) which I can provide if this is taken.

andyleejordan committed:
Add isearch-like bindings for case and style toggles

andyleejordan committed:
Add optional `arg` to `ctrlf--start` for minibuffer initial contents

This enables wrapper functions to be created with starting search terms.

andyleejordan committed:
Add optional `arg` to `ctrlf-forward` too

With the optional `arg`, the search will start with (or continue with)
it as the given search term. This lets wrapper functions supply initial
search terms.

andyleejordan committed:
Add and bind `ctrlf-forward-symbol-at-point`

This allows users that are accustomed to
`isearch-forward-symbol-at-point` semantics to use ctrlf with ease. The
`M-s .` mapping to start a symbol at point search (or populate an
existing search with symbol at point) now work, in addition to the
existing `M-n` option to use the first history element (which is also
the symbol at point).

andschwa commented:
TODO: Move to beginning of thing-at-point so search starts with thing instead of going to next thing if point is in the middle of thing.

andyleejordan committed:
Implement `ctrlf-occur` which acts much like `isearch-occur`

This is surprisingly straightforward because of the existing logic to
automatically translate the input based on its style into a regexp.

raxod502 reviewed (commented):
I agree, it makes more sense to reuse the `M-s` map as opposed to introduce something new like `C-o`. The only reason I added `C-o` was that I forgot about `M-s`.

Anyway, yes, having these bindings behave like they do in Isearch is a great improvement. The idea looks good to me.

raxod502 commented:
For symbol search, you could always use the regexp style and add delimiters.

> Move to beginning of thing-at-point so search starts with thing instead of going to next thing if point is in the middle of thing

Just a warning, this may be a bit tricky. Currently CTRLF just does a standard `search-forward` or `search-backward` from point. Does Isearch do something fancier?

andschwa commented:
Thanks @raxod502! I can split out the easy stuff (the bindings to existing functions, namely) while working on implementing the harder stuff (forward symbol search and occur). Any other tips/suggestions would be very welcome as I haven't worked a ton with search interfaces before (having to change point back from minibuffer to original buffer was one I learned doing this last night).

andyleejordan committed:
Add optional `pos` to `ctrlf--start` as starting point

This allows us to start the search from another position (say, the
beginning of the symbol at point), but retain the user's actual starting
position so that `ctrlf-cancel` behaves appropriately.

andyleejordan committed:
Improve handling of current point in `ctrlf-forward-symbol-at-point`

This now makes the search start at the beginning of the symbol at point,
so that if the point is within a symbol, it becomes the first search
result. Canceling the search drops the user back at the appropriate
position, the point at which they started.

andschwa commented:
> > Move to beginning of thing-at-point so search starts with thing instead of going to next thing if point is in the middle of thing
> 
> Just a warning, this may be a bit tricky. Currently CTRLF just does a standard `search-forward` or `search-backward` from point. Does Isearch do something fancier?

This ended up being a bit easier than I think we both expected. The function `isearch-symbol-at-point` cleverly uses `find-tag-default-bounds` to position the point at the beginning of the symbol at point. I made sure that that we retained the user's actual starting position so that `ctrlf-cancel` works appropriately.

I also deleted `ctrlf-backward-symbol-at-point` as there is no isearch equivalent, so that's inconsistent, and maintaining two is annoying.

andyleejordan commented on ctrlf.el:
```diff
@@ -99,7 +99,8 @@ lists with the following keys (all mandatory):
     ([remap isearch-forward-regexp ] . ctrlf-forward-regexp)
-    ([remap isearch-backward-regexp] . ctrlf-backward-regexp))
+    ([remap isearch-backward-regexp] . ctrlf-backward-regexp)
+    ([remap isearch-forward-symbol-at-point] . ctrlf-forward-symbol-at-point))
```
Do you want all of these realigned on the rightmost `.` now?

raxod502 replied:
It's up to your taste :)

andyleejordan replied:
Then nah cause cause it'll be a bigger diff and things will really slide to the right.

andyleejordan commented on ctrlf.el:
```diff
@@ -129,7 +130,15 @@ active in the minibuffer during a search."
+    ("M-s ."     . ctrlf-forward-symbol-at-point)
+    ;; Remaps `isearch-toggle-case-fold' which is in another map.
+    ("M-s c"     . ctrlf-toggle-case-fold-search)
     ("C-o c"     . ctrlf-toggle-case-fold-search)
```
Should we keep this and the other `C-o` binding?

raxod502 replied:
Sure, for now. We'll update the README to not document the old binding, and remove it eventually. But there's no reason to break backwards compatibility unless needed.

andyleejordan commented on ctrlf.el:
```diff
@@ -129,7 +130,15 @@ active in the minibuffer during a search."
     ([remap scroll-down-command]            . ctrlf-previous-page)
     ([remap recenter-top-bottom]            . ctrlf-recenter-top-bottom)
+    ;; Remaps `isearch-occur' which is in another map.
+    ("M-s o"     . ctrlf-occur)
```
I wish I could actually just remap these, but they are only temporarily bound when an isearch is active.

raxod502 replied:
Hmm... when I type `M-s C-h`, it says these commands are bound globally. Did I miss something?

andyleejordan replied:
So, `M-s o` is bound globally to `occur`. But in an `isearch`, `M-s o` is transiently rebound to `isearch-occur`. An `M-s C-h` here is misleading because even if you've started an isearch, it pulls up the global bindings.

I found `which-key` really helpful here because it showed the transient bindings, and I otherwise referred to the [manual](https://www.gnu.org/software/emacs/manual/html_node/emacs/Special-Isearch.html).

However, `ctrlf-forward-symbol-at-point` I goofed on. We're already defining it globally (as does isearch) and it doesn't need to be rebound again here. The rest of these, however, are only bound during `isearch`.

Note that like `M-s _` is slightly tricky because outside of the search it should be `ctrlf-forward-symbol` but inside a search it should be `ctrlf-toggle-symbol`.

andyleejordan commented on ctrlf.el:
```diff
@@ -753,8 +762,9 @@ And self-destruct this hook."
 
-(defun ctrlf--start ()
-  "Start CTRLF session assuming config vars are set up already."
+(defun ctrlf--start (&optional arg pos)
```
If you think of better variable names, let me know.

raxod502 replied:
What about INITIAL-CONTENTS and POSITION, following `read-from-minibuffer`?

andyleejordan replied:
_Much better._ Thanks!

andyleejordan commented on ctrlf.el:
```diff
@@ -884,18 +895,30 @@ change the search style if already in a search."
+         ;;
+         ;; TODO: Is there a better way to replace the minibuffer
+         ;; contents? Should we use `(with-current-buffer
+         ;; ctrlf--minibuffer...)' or is this safe?
```
I'm very new to manipulating the minibuffer contents like this. This was the best I could find based on research into `previous-history-element`. I've caught myself accidentally inserting over the entire current buffer when my code didn't properly leave the `with-current-buffer` statement before calling `ctrlf-forward`. This would be more of a crutch, and I think correct code should not require it.

andyleejordan commented on ctrlf.el:
```diff
@@ -884,18 +895,30 @@ change the search style if already in a search."
+         (t (ctrlf-next-match)))
       (setq ctrlf--backward-p nil)
-      (ctrlf--start))))
+      (ctrlf--start arg))))
```
I'm wondering if I should just pass `pos` through here too, but it doesn't make much other sense for `ctrlf-forward`. There are two main scenarios where we'd call `ctrlf-forward` from `ctrlf-forward-symbol-at-point`.

1. Hit `C-s`, enter no input, realize you want symbol under point, hit `M-s .` to start searching for it.
2. Be in the middle of a search, partially within a symbol, desire to search for the rest of the symbol, hit `M-s .` to fill it in. (Kind of like `C-w`, which I haven't implemented yet, but for the full symbol.)

andyleejordan commented on ctrlf.el:
```diff
@@ -884,18 +895,30 @@ change the search style if already in a search."
+      (ctrlf--start arg))))
 
-(defun ctrlf-backward (style &optional preserve)
+(defun ctrlf-backward (style &optional preserve arg)
```
I can revert these changes since I removed `ctrlf-backward-symbol-at-point`.

andyleejordan commented on ctrlf.el:
```diff
@@ -1002,6 +1033,48 @@ search, change back to regexp search."
+  "Search forward for symbol at point."
+  (interactive)
+  (let ((arg nil)
+        (pos nil))
```
Is there a pattern elisp pattern to follow for this? I need to get the values of `arg` and `pos` as they are within the `with-current-buffer` scope, but  I need to use them outside of it.

andyleejordan commented on ctrlf.el:
```diff
@@ -1002,6 +1033,48 @@ search, change back to regexp search."
+(defun ctrlf-occur ()
+  "Like `isearch-occur', but really basic!"
+  (interactive)
+  ;; TODO: Handle nil history, quoting etc., so much else...
```
Sorry, this isn't finished yet at all. It works for the very basic cases though!

andyleejordan committed:
Add symbol boundary constructs

We may want to make this optional. The former behavior, while less
consistent with isearch, was actually quite nice. It would search for
the symbol at point, but include strings of which it was the prefix, and
these could then be added to the search with `M-s .`.

andyleejordan committed:
Allow `ctrlf--message` to be used any time

In order to allow ctrlf commands to be able to display error messages
when the search does not start, `ctrlf--message` needs to defer to
`minibuffer-message` when the `ctrlf--minibuffer` is `nil`. This also
means we need to set it to `nil` when cleaning up.

This function had to be re-indented and a line break added.

andyleejordan committed:
Improve handling of no symbol at point

We now skip the search if there is no symbol at point, and display an
error message. This removes the workaround of sending an empty string to
`ctrlf-forward` when there was no symbol.

andschwa commented:
I have much improved `ctrlf-forward-symbol-at-point`, not to improve `ctrlf-occur` and perhaps implement `isearch-yank-word-or-char`.

andschwa commented:
@raxod502 Why is `(field-string (point-max))` used instead of `(minibuffer-contents)` in places? I saw the former after using the latter myself and am curious if there's a particular reason.

andschwa commented:
So, `ctrlf-occur` ended up being super easy to do. I think this is ready for review (except changelog updates). I'd love to add `C-w` too, but that one seems tricky to me, especially to get the isearch semantics right.

andyleejordan commented on ctrlf.el:
```diff
@@ -361,35 +370,38 @@ as a backport for Emacs 26 and earlier where signaling a message
 mess."
   ;; Some of this is borrowed from `minibuffer-message'.
-  (let ((string (apply #'format (concat " [" format "]") args)))
-    (put-text-property 0 (length string) 'face 'minibuffer-prompt string)
```
This looks like way more change than it actually is because of indentation. I just added handling of `ctrlf--minibuffer` being `nil` (and ensured it is such if a ctrlf search isn't active) such that the message is displayed consistently as if `minibuffer-message` were used directly. This lets us use `ctrlf--message` even if a search _wasn't_ started (like no symbol at point was found).

andyleejordan replied:
(Try `M-s .` with no symbol at point, then cancel and try `C-s` followed by `M-s .` for a demo of the logic.)

raxod502 replied:
Works great!

andyleejordan replied:
Awesome.

andyleejordan commented on ctrlf.el:
```diff
@@ -361,35 +370,38 @@ as a backport for Emacs 26 and earlier where signaling a message
+        ;; Setting REAR-ADVANCE:
+        ;; <https://github.com/raxod502/ctrlf/issues/4>
+        (let ((ol (make-overlay
+                   (point-max) (point-max) nil nil 'rear-advance)))
```
And then I had to add a newline to satisfy linters.

raxod502 replied:
Sounds good.

andyleejordan commented on ctrlf.el:
```diff
@@ -738,6 +750,7 @@ And self-destruct this hook."
 
 (defun ctrlf--minibuffer-exit-hook ()
   "Clean up CTRLF from minibuffer and self-destruct this hook."
+  (setq ctrlf--minibuffer nil)
```
This is that cleanup. Maybe this belongs in the `ctrlf--finalize`? I wasn't entirely sure. We need this because it's insufficient to check `ctrlf--active-p` if some logic is being performed before the start of a search.

raxod502 replied:
I think where you have it is fine. `ctrlf--finalize` is somewhat more of a hack, because it runs from `post-command-hook` and then removes itself right away. Ideally we wouldn't have to put anything there.

andyleejordan replied:
Good to know.

andyleejordan commented on ctrlf.el:
```diff
@@ -884,12 +899,20 @@ change the search style if already in a search."
-          (ctrlf-next-match))
+        (cond
+         ;; Continue search with `arg'.
+         ;; TODO: Is there a better way to replace the minibuffer contents?
```
Just calling out this question again 😂

andyleejordan replied:
If given `arg` I just want this function to replace any existing input with it.

raxod502 replied:
Maybe, but `delete-minibuffer-contents` followed by `insert` looks pretty great to me.

andyleejordan replied:
Cool!

andschwa commented:
You know, I had an idea last night that will add another isearch-like feature and simplify some things here. If I add a handler `symbol` we can have `M-s _` for `ctrlf-forward-symbol`, and defer to this function inside the symbol at point version.

andyleejordan committed:
Add `ctrlf-translate-symbol` translator

This simple refactor allows us to provide `ctrlf-forward-symbol` and
`ctrlf-backward-symbol`, and remap their isearch counterparts.

Furthermore, `ctrlf-forward-symbol-at-point` no longer has this logic
embedded in it.

andschwa commented:
Well, that was easy.

andyleejordan committed:
Add optional `pos` to `ctrlf-forward` as starting point

This is simply passed to `ctrlf--start` if the search is being started,
which allows wrapper functions to not deal with this special case.

andschwa commented:
Okay cool. I'm happy with this now. It's missing readme and changelog updates, but I'll do those after a review.

andyleejordan commented on ctrlf.el:
```diff
@@ -871,11 +896,12 @@ Wrap around if necessary."
 ;;;;; Main commands
 
-(defun ctrlf-forward (style &optional preserve)
+(defun ctrlf-forward (style &optional preserve arg pos)
```
It turned out it _was_ better to pass both `arg` and `pos` here. Since we pass `pos` straight to `ctrlf--start`, wrapper functions don't have to deal with the edge cases of being in or out of a search, which keeps this logic in exactly one place.

raxod502 replied:
Okay!

andyleejordan commented on ctrlf.el:
```diff
@@ -212,6 +226,12 @@ See `ctrlf-split-fuzzy' for how INPUT is split into subinputs.
 The subinputs are joined with \".*\"."
   (string-join (ctrlf-split-fuzzy input) ".*"))
 
+(defun ctrlf-translate-symbol (input)
```
Super nice side effect of this is that the symbol boundary constructs don't appear in the user's input, making it exactly like `isearch-forward-symbol`. Furthermore, we automatically get correct behavior with `ctrlf-occur`, too.

raxod502 replied:
That looks great.

andyleejordan replied:
Thanks! It really cleaned up when I did this.

raxod502 commented on ctrlf.el:
```diff
@@ -73,6 +73,9 @@ Otherwise, the match count is only shown in the minibuffer."
+                             :case-fold ctrlf-no-uppercase-regexp-p))
+    (symbol       . (:prompt "symbol"
+                             :translator ctrlf-translate-symbol
                              :case-fold ctrlf-no-uppercase-regexp-p)))
```
```suggestion
                             :case-fold ctrlf-no-uppercase-literal-p)))
```

andyleejordan replied:
Good catch, thanks!

andschwa commented:
For what it's worth, I've been using this the last few days and it's working great!

andyleejordan committed:
Add `ctrlf-toggle-regexp` and `ctrlf-toggle-symbol`

So that we can correctly bind `M-s r` and `M-s _` when in a CTRLF
session.

andyleejordan committed:
Update readme and changelog

andschwa commented:
I've rebased, updated the readme and changelog, and on closer inspection, added `ctrlf-toggle-regexp` and `ctrlf-toggle-symbol` with matching bindings to get as close to [Isearch](https://www.gnu.org/software/emacs/manual/html_node/emacs/Special-Isearch.html) as possible.

andschwa commented:
I want to add `isearch-query-replace`/`M-%`, `isearch-yank-word-or-char`/`C-w`, and `isearch-yank-line`/`M-s C-e` at some point too.

raxod502 commented on ctrlf.el:
```diff
@@ -927,6 +967,18 @@ Interactively, select from the list of all defined styles."
+  (interactive)
+  (setq ctrlf--style
+        (if (eq ctrlf--style 'symbol) 'literal 'symbol)))
+
```
Having separate toggles for these is a great idea. That is faster to use than always having to select from a list.

andyleejordan replied:
Hey I didn't think of it! Thank you long-forgotten Emacs/Isearch developer 😆

raxod502 commented on ctrlf.el:
```diff
@@ -1015,6 +1067,61 @@ search, change back to regexp search."
+      ;; how else it could be implemented.
+      (let ((start (car (find-tag-default-bounds)))) ; Find start of symbol.
+        (when (and start (< start (point))) ; `start' can be nil.
+          (setq ctrlf--current-starting-point start) ; For `ctrlf-forward'.
```
Hmmm... I was going to say we should avoid using internal functions in these wrappers, but then I looked at the other ones and they do that too. I guess at some point I forgot that I was trying to do that. Room for future improvement!

andyleejordan replied:
Oh yeah, always room for improvement 😂

raxod502 commented on ctrlf.el:
```diff
@@ -1015,6 +1067,61 @@ search, change back to regexp search."
+    (unless skip-search
+      ;; NOTE: We pass `pos' so that `ctrlf-cancel' returns to the
+      ;; original position, and not the start of the symbol at point.
+      (ctrlf-forward 'symbol nil arg pos))))
```
Would it work if in `ctrlf--start` you set `ctrlf--current-starting-point` to `pos` instead of `ctrlf--starting-point`? Then you might not have to move point in this function, and hence not have to save its original value.

andyleejordan replied:
I feel like I tried that, but let me try it out again.

andyleejordan replied:
Ah ha, that worked nicely!

raxod502 commented on CHANGELOG.md:
```diff
@@ -13,13 +13,18 @@ The format is based on [Keep a Changelog].
+* More Isearch compatibility was added in [#48]. Additions include:
+  `ctrlf-occur`, `ctrlf-forward-symbol`,
+  `ctrlf-forward-symbol-at-point`, `ctrlf-toggle-regexp`,
+  `ctrlf-toggle-symbol`, and matching Isearch bindings for those.
```
Perhaps this would work better in the Features section (can be added just above Enhancements)?

raxod502 reviewed (commented):
Sorry for the delay. This is a really great contribution, thank you for putting so much thought into it! I have some thoughts, let me know what you think. All minor things.

andyleejordan committed:
Don't `ctrlf-forward-symbol-at-point` twice

We're already defining it globally (as Isearch does) and it doesn't need
to be rebound again here. The rest of these, however, are only bound
during a search.

andschwa commented:
Rebased 😂 I just need to try the different approach to `pos`, and then rename the vars.

andyleejordan committed:
Simplify setting the current starting point in symbol search

We can avoid moving `point` by setting the _current_ starting point
instead of the "starting point," reducing complexity.

andyleejordan committed:
Rename `arg` and `pos` to `initial-contents` and `position`

andschwa commented:
@raxod502 I think you can merge this now when you see fit. Let me know if there's anything else!

andschwa commented:
@raxod502 were you waiting on anything else here?

raxod502 commented:
Nope, I've just been very busy with other life things and have 40 threads backlogged at the moment. I will definitely get back to this as soon as possible. Sorry for the inconvenience!

andschwa commented:
Fair! I definitely get that. Take care!

raxod502 reviewed (approved):
Great! Thanks for your patience!
